feat: Open telemetry integration and span Id fix for nodejs logging library

cindy-peng opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-logging/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1490  🦕

This change is to add OpenTelemetry support for nodejs-logging library and fix trace-span correlation issue.
More context about this change: https://b.corp.google.com/issues/270985947

Here are the overall and nodejs-logging design doc for this change:
[go/logging-otel-span-support](http://goto.google.com/logging-otel-span-support)
[go/logging-otel-span-support-nodejs](http://go/logging-otel-span-support-nodejs)

Commit:
feat: OpenTelemetry integration

Commit:
feat: add otel system testing with cloud tracing

Commit:
Add unit testing for otel context

Commit:
Add parseOtelContext test

Commit:
testing entry metadata population

Commit:
Add system test suites

Commit:
remove only mocha test keywords

Commit:
Fix lint errors

Commit:
fix lint issues

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

Commit:
Add span end method in system tests

cindy-peng committed:
Merge branch 'main' into cindy/spanidfix

Commit:
fix gts check

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'cindy/spanidfix' of https://github.com/googleapis/nodejs-logging into cindy/spanidfix

gcf-owl-bot[bot] committed:
Merge branch 'cindy/spanidfix' of https://github.com/googleapis/nodejs-logging into cindy/spanidfix

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng reviewed (commented):
Could you put the changes in `protos/` in a separate PR? Not sure how related they are to the current changes.

daniel-sanche commented on package.json:
```diff
@@ -93,7 +95,13 @@
     "webpack": "^5.0.0",
-    "webpack-cli": "^5.0.0"
+    "webpack-cli": "^5.0.0",
+    "@opentelemetry/sdk-trace-node": "^1.23.0",
+    "@opentelemetry/sdk-node": "^0.50.0",
+    "@opentelemetry/semantic-conventions": "^1.23.0",
+    "@opentelemetry/sdk-trace-base": "^1.23.0",
+    "@opentelemetry/resources": "^1.23.0",
+    "@google-cloud/opentelemetry-cloud-trace-exporter": "^2.1.0"
```
+1, my understanding is that we just needed a hard dependency on the api for this feature, not the sdk?

cindy-peng replied:
Yes @gkevinzheng @daniel-sanche  these are only for system tests and unit tests and they are already under `devDependencies`. Customers won't need to install these packages.

daniel-sanche replied:
Oh perfect, I missed that!

gkevinzheng replied:
Same here, missed that.

daniel-sanche commented on package.json:
```diff
@@ -54,8 +56,8 @@
-    "google-auth-library": "^9.0.0",
-    "google-gax": "^4.0.3",
+    "google-auth-library": "^9.3.0",
+    "google-gax": "^4.3.0",
```
Is this version bump required?

cindy-peng replied:
It doesn't look like so, reverted:)

daniel-sanche reviewed (commented):
Mostly LGTM, but see if you can reduce those dependencies

cindy-peng commented on src/utils/context.ts:
```diff
@@ -149,6 +156,31 @@ function makeCloudTraceHeader(): string {
+  const spanContext = trace.getActiveSpan()?.spanContext();
+  if (spanContext !== undefined && isSpanContextValid(spanContext)) {
+    const otelSpanContext = {
+      // trace: "projects/cindy-cloud-sdk-test/traces/"+spanContext?.traceId,
```
Yea I was trying to remove these as well, but it looks like owlbot committed this automatically to this PR: https://github.com/googleapis/nodejs-logging/pull/1497/commits/c25466975361fabd0b9434be6e50f8d624fffbee

cindy-peng replied:
Done.

cindy-peng commented:
> Could you put the changes in `protos/` in a separate PR? Not sure how related they are to the current changes.

Yea I was trying to remove these as well, but it looks like owlbot committed this automatically to this PR after I removed them: https://github.com/googleapis/nodejs-logging/commit/c25466975361fabd0b9434be6e50f8d624fffbee

Commit:
Remove comments

cindy-peng commented on test/entry.ts:
```diff
@@ -19,6 +19,12 @@ import * as proxyquire from 'proxyquire';
+import {InMemorySpanExporter} from '@opentelemetry/sdk-trace-base';
+import {trace} from '@opentelemetry/api';
+import {Resource} from '@opentelemetry/resources';
+//const {Resource} = require('@opentelemetry/resources');
```
Done.

cindy-peng commented on system-test/logging.ts:
```diff
@@ -29,6 +29,18 @@ const http2spy = require('http2spy');
+// } = require('@opentelemetry/semantic-conventions');
+// const {
+//   TraceExporter,
+// } = require('@google-cloud/opentelemetry-cloud-trace-exporter');
```
Done.

Commit:
move otel packages to peer dependencies

Commit:
move otel packages to peer dependencies

cindy-peng commented on package.json:
```diff
@@ -45,6 +45,8 @@
     "postpublish": "./.kokoro/publish-min.sh"
   },
   "dependencies": {
+    "@opentelemetry/api": "^1.7.0",
```
Good catch! Done.

cindy-peng replied:
Hi @gkevinzheng, 
It looks like our kokoro tests ran on NPM 6.14.18 which doesn't support auto-installment for peer-dependencies: https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/#

This is causing [system test module not found test failure](https://btx.cloud.google.com/invocations/1f915947-5739-426a-bc6f-36ce8878fb85/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Fpresubmit%2Fgoogleapis%2Fnodejs-logging%2Fnode14%2Fsystem-test;config=default/log).

I have posted in JS-Team chat but it doesn't look like there is any solution so far. There is a plan to upgrade kokoro infra to Node18 but not sure when it will be done.

We will probably need to move this back to `Dependencies` instead of `peerDependencies` due to this limitation.

cindy-peng replied:
Due to the limitation of low NPM version in kokoro system test infra, JS language team suggested us to import OpenTelemetry/api and re-export it for now to avoid test failure and guarantee compatibility : https://chat.google.com/room/AAAAW_Lho4o/ljx6XogZJMI
Once NPM version in Kokoro infra has upgraded, I will change this to peer dependency again.

Commit:
fix system test

gkevinzheng reviewed (approved):
LGTM

cindy-peng committed:
Merge branch 'main' into cindy/spanidfix

Commit:
add bitmask for tracesampled flag

Commit:
fix pack-n-play testing

cindy-peng commented:
> > Could you put the changes in `protos/` in a separate PR? Not sure how related they are to the current changes.
> 
> Yea I was trying to remove these as well, but it looks like owlbot committed this automatically to this PR after I removed them: [c254669](https://github.com/googleapis/nodejs-logging/commit/c25466975361fabd0b9434be6e50f8d624fffbee)

Removed proto changes.

Commit:
revert pack-n-play sample configuration

Commit:
Make otel api as dependency due to low npm version in kokoro infra

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

aabmass commented on package.json:
```diff
@@ -59,7 +59,8 @@
     "stream-events": "^1.0.5",
-    "uuid": "^9.0.0"
+    "uuid": "^9.0.0",
+    "@opentelemetry/api": "^1.7.0"
```
Would peerDependency work better here?

cindy-peng replied:
Yea there was some back and forth discussions here: https://github.com/googleapis/nodejs-logging/pull/1497#discussion_r1578061360

Basically system test would fail for now if it is imported as peer dependency since our kokoro runs on NPM 6, which doesn't install peer dependency automatically. JS language team suggested us to import OpenTelemetry/api and re-export it for now to avoid test failure and guarantee compatibility : https://chat.google.com/room/AAAAW_Lho4o/ljx6XogZJMI
Once NPM version in Kokoro infra has upgraded, I can change this to peer dependency again.

aabmass commented on src/utils/context.ts:
```diff
@@ -149,6 +156,31 @@ function makeCloudTraceHeader(): string {
+      traceSampled: (spanContext.traceFlags & FLAG_SAMPLED) !== 0,
+    };
+
+    const return_context = toCloudTraceContext(otelSpanContext, projectId);
+    return return_context;
```
nit could just return directly instead of the assignment

cindy-peng replied:
Done.

aabmass reviewed (approved):
Didn't look at the tests too closely but LGTM from an OTel perspective

Commit:
fix comments to return value directy

cindy-peng committed:
Merge branch 'main' into cindy/spanidfix

cindy-peng committed:
Merge branch 'main' into cindy/spanidfix
